src/compute/src/render/reduce.rs

```diff
@@ -1174,14 +1197,20 @@ where
                     // of the multiplicity of the final result in the input, we only want to have one copy
                     // in the output.
                     target.push((R::ok(row_builder.clone()), -1));
-                    target.extend(
-                        source
-                            .iter()
-                            .map(|(values, cnt)| (R::ok((*values).clone()), *cnt)),
-                    );
+                    for (values, cnt) in source.iter() {
+                        row_packer = row_builder.packer();
+                        row_packer.extend(*values);
+                        target.push((R::ok(row_builder.clone()), *cnt));
+                    }
```

Comment:
The change to imperative style here is to avoid a closure and be able to continue using the `row_packer` / `row_builder`.

Reply:
We need to make sure `target` has enough capacity, maybe we can ensure this?

Reply:
I added a `reserve` call with the number of `push` calls we'll do to `target`. IIUC that would get us a similarly sized allocation as using `extend`. As with `extend`, though, the allocation is not guaranteed to be tight, but we avoid reaching out to the allocator too many times.